File: wwan_keep_alive.pl
#!/usr/bin/perl

use strict;
use warnings;
use threads;
use threads::shared;
use IO::Handle;

my $sleep_sec = 3;

sub println {
    my $string = shift;
    my $dt     = time();
    print("[$dt]: $string\n");
}

sub write_reboot {
    my $name = shift;
    open( my $reboots_file, '>>', "/home/pi/reboots.log" );
    my $t = time();
    print $reboots_file "$t - $name restarted.\n";
    close $reboots_file;
}

sub connect_data {
    println("Entering connect_data()...");
    while (1) {
        println("Connecting Data...");
        sleep $sleep_sec;
        system("qmicli -d /dev/cdc-wdm0 --dms-set-operating-mode='online'");
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        system("ip link set wwan0 down");
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        system("echo 'Y' | tee /sys/class/net/wwan0/qmi/raw_ip");
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        system("ip link set wwan0 up");
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        system(
"qmicli -p -d /dev/cdc-wdm0 --device-open-net='net-raw-ip|net-no-qos-header' --wds-start-network='ip-type=4' --client-no-release-cid"
        );
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        system("timeout 20 udhcpc -i wwan0");
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        sleep $sleep_sec;
        sleep $sleep_sec;

        # try a ping
        system("ping 8.8.8.8 -I wwan0 -c1 -W2");
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        println("Connected Data!");
        last;
    }
}

sub connect_gps {
    println("Entering connect_gps()...");
    while (1) {
        println("Connecting GPS...");
        println(
"Checking if /dev/ttyUSB1 exists. If it doesn't, we need to reboot the pi."
        );
        if ( !-e "/dev/ttyUSB1" ) {
            println(
"/dev/ttyUSB1 does not exist. I would have done a reboot just now."
            );

            # write_reboot("system");
            # system("reboot");
            next;
        }
        sleep $sleep_sec;
        system(
            "stty -F /dev/ttyS0 115200 raw -echo -echoe -echok -echoctl -echoke"
        );
        next if $? >> 8 != 0;
        sleep $sleep_sec;
        my $filename = "/dev/ttyUSB2";
        open( my $tty, '>>', $filename ) or die "Could not open file $filename";
        print $tty "AT+CGPS=0\r";
        $tty->autoflush;
        sleep $sleep_sec;
        print $tty "AT+CGPSDEL\r";
        $tty->autoflush;
        sleep $sleep_sec;
        print $tty "AT+CGPSAUTO=1\r";
        $tty->autoflush;
        sleep $sleep_sec;
        print $tty "AT+CGPS=1\r";
        $tty->autoflush;
        sleep $sleep_sec;
        system("systemctl restart gpsd");
        next if $? >> 8 != 0;
        println("Connected GPS!");
        close $tty;
        last;
    }
}

sub handle_data {

    connect_data();

    # make sure ping works
    while (1) {
        my $count = 0;
        while ( $count < 10 ) {
            sleep 1;
            system("ping -c 1 -I wwan0 8.8.8.8");

            # increment count if the ping failed, reset if passed
            $? >> 8 == 0 ? $count = 0 : $count++;
            println("data count = $count");
        }

        # reconnect data
        write_reboot("data");
        connect_data();
    }
}

sub handle_gps {
    connect_gps();

    # make sure gps works
    while (1) {
        my $count = 0;
        while ( $count < 5 ) {
            sleep 1;
            my $gpspipe_output = `timeout 10 gpspipe -w`;

            # increment count if no TPV data, reset if passed
            index( $gpspipe_output, "TPV" ) != -1 ? $count = 0 : $count++;
            println("gps count = $count");
        }

        # reconnect gps
        write_reboot("gps");
        connect_gps();
    }
}

## MAIN ##

sleep $sleep_sec;
my $data_thread = threads->create('handle_data');
my $gps_thread  = threads->create('handle_gps');
$data_thread->join();
$gps_thread->join();

File: wwan_keep_alive.py
#!/usr/bin/python3 -u

import subprocess
import logging
import threading
import time
import os
import serial

logging.basicConfig(
    format="%(asctime)s - %(message)s", datefmt="%d-%b-%y %H:%M:%S", level=logging.INFO
)


DEFAULT_SLEEP_SECONDS = 3


def sleep(seconds=DEFAULT_SLEEP_SECONDS):
    logging.info(f"I should be sleeping for {seconds} seconds")
    time.sleep(seconds)


def system(command: str) -> bool:
    return os.system(command) == 0


def system_and_sleep(command: str) -> bool:
    logging.info(f"Running command: {command}")
    code = system(command)
    sleep()
    return code


def write_reboot(name: str):
    logging.info("Writing entry to reboots.log...")
    with open("a", "/home/pi/reboots.log") as f:
        now = int(time.time)
        f.write(f"{now} - {str} restarted.")


def connect_data():
    logging.info("Entering connect_data()...")
    while True:
        logging.info("Connecting Data...")
        if not system_and_sleep(
            "qmicli -d /dev/cdc-wdm0 --dms-set-operating-mode='online'"
        ):
            continue
        if not system_and_sleep("ip link set wwan0 down"):
            continue
        if not system_and_sleep("echo 'Y' | tee /sys/class/net/wwan0/qmi/raw_ip"):
            continue
        if not system_and_sleep("ip link set wwan0 up"):
            continue
        if not system_and_sleep(
            "qmicli -p -d /dev/cdc-wdm0 --device-open-net='net-raw-ip|net-no-qos-header' --wds-start-network='ip-type=4' --client-no-release-cid"
        ):
            continue
        if not system_and_sleep("timeout 20 udhcpc -i wwan0"):
            continue
        sleep(10)
        if not system_and_sleep("ping 8.8.8.8 -I wwan0 -c1 -W2"):
            continue
        logging.info("Connected Data!")
        break


def connect_gps():
    logging.info("Entering connect_gps()...")
    while True:
        logging.info("Connecting GPS...")
        logging.info(
            "Checking if /dev/ttyUSB1 exists. If it doesn't, we need to reboot the pi."
        )
        if not os.path.exists("/dev/ttyUSB1"):
            logging.info("/dev/ttyUSB1 does not exist. Rebooting the system")
            # write_reboot("system")
            # if not system_and_sleep("reboot"):
            #     logging.info("Could not reboot the system!!!")
        else:
            sleep()
            if not system_and_sleep(
                "stty -F /dev/ttyS0 115200 raw -echo -echoe -echok -echoctl -echoke"
            ):
                continue
            ser = serial.Serial("/dev/ttyUSB2", 115200)
            ser.write("AT+CGPS=0\r")
            sleep()
            ser.write("AT+CGPSDEL\r")
            sleep()
            ser.write("AT+CGPSAUTO=1\r")
            sleep()
            ser.write("AT+CGPS=1\r")
            sleep()
            if not system_and_sleep("systemctl restart gpsd"):
                continue
            logging.info("Connected GPS!")
            break


def handle_data():
    connect_data()
    while True:
        count = 0
        while count < 10:
            sleep(1)
            if system("ping -c 1 -I wwan0 1.1.1.1"):
                count = 0
            else:
                count += 1
        # reconnect data
        write_reboot("data")
        connect_data()


def handle_gps():
    connect_gps()
    while True:
        count = 0
        while count < 5:
            sleep(1)
            output = subprocess.check_output("timeout 10 gpspipe -w")
            if "TPV" in output:
                count = 0
            else:
                count += 1
            logging.info(f"gps count = {count}")
        # reconnect gps
        write_reboot("gps")
        connect_gps()


def main():
    sleep()
    data_thread = threading.Thread(target=handle_data)
    data_thread.start()
    gps_thread = threading.Thread(target=handle_gps)
    gps_thread.start()
    data_thread.join()
    gps_thread.join()


main()
